pop-api/examples/fungibles/tests.rs

```diff
@@ -0,0 +1,433 @@
+	let _ = env_logger::try_init();
+	// Deploy a new contract.
+	let contract = deploy_with_new_constructor(
+		&mut session,
+		BundleProvider::local()?,
+		TOKEN_ID,
+		TOKEN_MIN_BALANCE,
+	)?;
+
+	const AMOUNT: Balance = TOKEN_MIN_BALANCE * 4;
+	const TRANSFERRED: Balance = TOKEN_MIN_BALANCE;
+	// Mint tokens.
+	assert_ok!(session.sandbox().mint_into(&TOKEN_ID, &contract.clone(), AMOUNT));
+	assert_ok!(session.sandbox().mint_into(&TOKEN_ID, &BOB, AMOUNT));
+	// Check balance of accounts.
+	assert_eq!(session.sandbox().balance_of(&TOKEN_ID, &contract.clone()), AMOUNT);
+	assert_eq!(session.sandbox().balance_of(&TOKEN_ID, &BOB), AMOUNT);
+
+	session.set_actor(contract.clone());
+	// No-op if `value` is zero.
+	assert_ok!(transfer(&mut session, ALICE, 0));
+	assert_eq!(last_contract_event(&session), None); // No event emitted.
+
+	// No-op if the caller and `to` is the same address, returns success and no events are emitted.
+	assert_ok!(transfer(&mut session, contract.clone(), TRANSFERRED));
+	assert_eq!(last_contract_event(&session), None); // No event emitted.
+
+	// Failed with `InsufficientBalance`.
+	let data = serde_json::to_string::<[u8; 0]>(&[]).unwrap();
+	expect_call_reverted(
+		&mut session,
+		TRANSFER,
+		vec![BOB.to_string(), (AMOUNT + 1).to_string(), data],
+		PSP22Error::InsufficientBalance,
+	);
+	// Make sure balance is not changed.
+	assert_eq!(session.sandbox().balance_of(&TOKEN_ID, &contract), AMOUNT);
+
+	// Successfully transfer tokens from `contract` to `account`.
+	assert_ok!(transfer(&mut session, BOB, TRANSFERRED));
+	let expected = Transfer {
+		from: Some(account_id_from_slice(contract.clone().as_ref())),
+		to: Some(account_id_from_slice(BOB.as_ref())),
+		value: TRANSFERRED,
+	}
+	.encode();
+	// A `Transfer` event is emitted.
+	assert_eq!(last_contract_event(&session).unwrap(), expected.as_slice());
+
+	assert_eq!(session.sandbox().balance_of(&TOKEN_ID, &contract), AMOUNT - TRANSFERRED);
+	assert_eq!(session.sandbox().balance_of(&TOKEN_ID, &BOB), AMOUNT + TRANSFERRED);
+	Ok(())
+}
+
+#[drink::test(sandbox = Sandbox)]
+fn transfer_from_works(mut session: Session) -> Result<(), Box<dyn std::error::Error>> {
+	let _ = env_logger::try_init();
+	// Deploy a new contract.
+	let contract = deploy_with_new_constructor(
+		&mut session,
+		BundleProvider::local()?,
+		TOKEN_ID,
+		TOKEN_MIN_BALANCE,
+	)?;
+
+	const AMOUNT: Balance = TOKEN_MIN_BALANCE * 4;
+	// Mint tokens.
+	assert_ok!(session.sandbox().mint_into(&TOKEN_ID, &ALICE, AMOUNT));
+	// Approve `contract` to spend `ALICE` tokens.
+	assert_ok!(session.sandbox().approve(&TOKEN_ID, &ALICE, &contract.clone(), AMOUNT * 2));
+	assert_eq!(session.sandbox().allowance(&TOKEN_ID, &ALICE, &contract.clone()), AMOUNT * 2);
+	// Check balance of accounts.
+	assert_eq!(session.sandbox().balance_of(&TOKEN_ID, &contract), 0);
+	assert_eq!(session.sandbox().balance_of(&TOKEN_ID, &ALICE), AMOUNT);
+	assert_eq!(session.sandbox().balance_of(&TOKEN_ID, &BOB), 0);
```

Comment:
No tests for code form other crates, it pollutes the tests. Tests should test the code that is written here. Adding these tests draw attention away of what actually needs attention and proper testing. Also more code, more maintenance, longer to go through the code when you have to get back to it after a while.